lib/internal/process/promises.js

```diff
@@ -41,88 +41,182 @@ const { isErrorStackTraceLimitWritable } = require('internal/errors');
 // *Must* match Environment::TickInfo::Fields in src/env.h.
 const kHasRejectionToWarn = 1;
 
-const maybeUnhandledPromises = new SafeWeakMap();
-const pendingUnhandledRejections = [];
-const asyncHandledRejections = [];
-let lastPromiseId = 0;
+// By default true because in cases where process is not a global
+// it is not possible to determine if the user has added a listener
+// to the process object.
+let hasMultipleResolvesListener = true;
 
-// --unhandled-rejections=none:
-// Emit 'unhandledRejection', but do not emit any warning.
-const kIgnoreUnhandledRejections = 0;
+if (process.on) {
+  hasMultipleResolvesListener = process.listenerCount('multipleResolves') !== 0;
 
-// --unhandled-rejections=warn:
-// Emit 'unhandledRejection', then emit 'UnhandledPromiseRejectionWarning'.
-const kAlwaysWarnUnhandledRejections = 1;
+  process.on('newListener', (eventName) => {
+    if (eventName === 'multipleResolves') {
+      hasMultipleResolvesListener = true;
+      resolveErrorResolve = resolveErrorResolveHandler;
+      resolveErrorReject = resolveErrorRejectHandler;
+    }
+  });
 
-// --unhandled-rejections=strict:
-// Emit 'uncaughtException'. If it's not handled, print the error to stderr
-// and exit the process.
-// Otherwise, emit 'unhandledRejection'. If 'unhandledRejection' is not
-// handled, emit 'UnhandledPromiseRejectionWarning'.
-const kStrictUnhandledRejections = 2;
+  process.on('removeListener', (eventName) => {
+    if (eventName === 'multipleResolves') {
+      hasMultipleResolvesListener = process.listenerCount('multipleResolves') !== 0;
+      if (!hasMultipleResolvesListener) {
+        resolveErrorResolve = noop;
+        resolveErrorReject = noop;
+      }
```

Comment:
If I read the code correct, this is not needed. The `hasMultipleResolvesListener` variable already prevents the resolvers from being called. Changing them to noop should therefore never be required, right?

Reply:
I will have a look later.

Reply:
done